### PR TITLE
Update kinst.cpp

### PR DESCRIPTION
**Pull request by pdaderko**
Emulates Killer Instinct 2 Upgrade Kit IDE protection PCB, adding support for kinst2uk set.

### CODE REVIEW

**Comment by pdaderko** [comment]
I added emulation of the Killer Instinct 2 Upgrade Kit CPLD based on reverse engineering the original device with a logic analyzer and custom test ROMs.  The idea of the protection is that it sits on the IDE bus as a slave device, and returns data when selected, based on previous writes.

The protection scheme has two 4-bit registers and a 4-bit counter. Writing to the device with D6 high stores the lower 4 bits (head number, according to the IDE registers) in an internal register. Writing to the device with D6 low stores the lower 4 bits in another internal register, and also increments a 4-bit counter. Reading the device returns the value of the counter XOR register 1 XOR register 2. Reading doesn't change any data.

Since the device is an IDE slave, it doesn't interfere with the dedicated ROMs, which only interface to the IDE master.  On real hardware, I've tested that you can run the dedicated ROMs with the upgrade kit PCB installed without issues.  So, I didn't make the protection code conditional on running with kinst2uk only.  Feel free to change that, or variable names, etc. in order to stay consistent with the rest of the MAME code.

I added some details in this forum post as well: https://www.arcade-projects.com/threads/killer-instinct-dual-board.5929/page-4#post-289824 , and a spreadsheet of the PCB/CPLD pin mappings to this post: https://www.arcade-projects.com/threads/killer-instinct-dual-board.5929/page-3#post-288973 .

**Comment by smf-** [comment]
It would be better to have this as slot device, you can implement a
class locally in the kinst driver.

Then it would be easy to have it hooked up only for the game that uses it.

**Comment by happppp** [comment]
A slot device for this is like playing on hard mode. I'd just add a subclass and make the ide_r/w functions virtual.
Other than that, small cleanups like adding spaces for readability (x == 0 && y == 1, instead of x==0&&y==1), indentation tabs(instead of 2*space), and the m_ prefix we use for class members.

**Comment by Osso13** [comment]
Also remember that the driver supports saving, so remember to save the new variables, too.

**Comment by smf-** [comment]
What?

**Comment by pdaderko** [comment]
Thanks for the comments.  I can definitely make the simple changes, but maybe someone can help with a bit of hand-holding on some of the specifics as 1) I'm not very familiar with the MAME source (I've recompiled and made minor mods several times over the past 15 or so years, usually to help with ROM hacks or other reverse engineering, but never adding much real functionality) and 2) I'm an EE that can code, but my daily coding is C on embedded 8 to 32 bit MCUs... so admittedly, my OOP is a bit weak (I'm not a computer scientist).

> I'd just add a subclass and make the ide_r/w functions virtual.
So, a subclass of kinst_state (we'll call it kinst2uk_state), overriding the ide_r and ide_w functions... then in the kinst2uk GAME() line, use kinst2uk_state?

> Also remember that the driver supports saving, so remember to save the new variables, too.
Is this described somewhere?  It seems that the MAMEdev wiki is down... I found a cached page talking about it, but it looks out of date ( https://wiki.mamedev.org/index.php/Save_State_Fundamentals ), unless I'm just not looking in the right spot to find anything described on that page.

Side note - I find there's a lot of learning curve because of all the MAME specific references, and I don't know an easy way to see what the code actually does (like "go to definition" like I'd do in an Eclipse based project).  Is there auto-generated code documentation somewhere that I'm missing?  Some things are easy to find by searching, but others are used so frequently that finding the definition based on a keyword search is nearly impossible (e.g. the definition of GAME() ).

Anyway - mostly, I just want to make sure the details of this protection hardware gets documented - so if you can bear with me while I get up to speed, that's great... or if it's so easy that you'd rather just fix the code yourself, feel free. ;-)

**Comment by Osso13** [comment]
Every variable which gets assigned values during live emulation (as opposed to only at start up) needs to be saved.
In your case just add save_item(NAME(your_variable_here)); in machine_start().
As for documentation, admittedly MAME isn't great yet. But most framework stuff is in src/emu, so if you limit yor search for definitions there you'll find them quite easily.

**Comment by happppp** [comment]
For me if I don't know something, the MAME source is the documentation.
The subclass thing, something like this:
`
remove "private:" from main class

class kinst2uk_state : public kinst_state
{
public:
	kinst2uk_state(const machine_config &mconfig, device_type type, const char *tag) :
		kinst_state(mconfig, type, tag)
	{ }

protected:
	virtual void machine_start() override;
	virtual ide overrides

private:
	m_prot_stuff vars
};


under kinst machine_start

void kinst2uk_state::machine_start()
{
	kinst_state::machine_start();
	
	// savestates here
}


make a comment header for example "kinst2uk protection handling"
You can also write tech info there if you want. Like the stuff you wrote in yesterday's comment.

void kinst2uk_state::ide_w(offs_t offset, uint32_t data, uint32_t mem_mask)
{
	// protection stuff here

	kinst_state::ide_w(offset, data, mem_mask);
}
`

File: src/mame/drivers/kinst.cpp [hunk 1]
```diff
@@ -374,12 +382,31 @@ INTERRUPT_GEN_MEMBER(kinst_state::irq0_start)
...
+    else /* D6 low */
+    {
+      prot_sel=0x01; /* select IDE slave (D6 low) */
+      prot_regb=data&0x0F; /* store lower 4 bits in register b */
+      prot_cnt=(prot_cnt+1)&0x0F; /* every write here increments counter */
+    }
+  }
+  else if ((offset==0x0c)&&(!(data&0x10))) /* write to IDE master */
+    prot_sel=0; /* deselect IDE slave */
 	m_ata->cs0_w(offset / 2, data, mem_mask);
```
Comment (cuavas) [inline src/mame/drivers/kinst.cpp:410]: Please follow out coding conventions.  Use one tab indent per level of scope, use lowercase hex digits consistently, and don’t code like your spacebar is broken, and don’t code like your space bar is broken.

Also, the CPLD is only present for the upgrade kit, and hence shouldn’t be emulated all the time.  It also affects only a single offset.

* Make separate read/write functions for handing the offset affected by the CPLD.  You can call through to the regular `ide_r` and `ide_w` from your read/write functions when the CPLD isn’t intercepting the access.
* Make a derived address map function for the upgrade kit, and install your handlers over the top at the offset they affect.
* Make a derived machine configuration function for the upgrade kit and set your modified address map.
* Use this machine configuration function in the `GAME` macro for the upgrade kit.

File: src/mame/drivers/kinst.cpp [hunk 2]
```diff
@@ -305,6 +307,12 @@ void kinst_state::machine_reset()
 
 	/* set a safe base location for video */
 	m_video_base = &m_rambase[0x30000/4];
+	
+	/* initialize kinst2uk protection registers */
+	prot_sel=0;
+	prot_cnt=0;
+	prot_rega=0;
+	prot_regb=0;
```
Comment (cuavas) [inline src/mame/drivers/kinst.cpp:315]: Is the CPLD actually connected to the machine reset line?  If not, these should only be reset in `machine_start`.

File: src/mame/drivers/kinst.cpp [hunk 3]
```diff
@@ -227,6 +227,8 @@ class kinst_state : public driver_device
 	uint32_t *m_video_base;
 	const uint8_t *m_control_map;
 	emu_timer *m_irq0_stop_timer;
+	
+	uint8_t prot_sel, prot_cnt, prot_rega, prot_regb;
```
Comment (cuavas) [inline src/mame/drivers/kinst.cpp:231]: Please follow the coding style of the class and use the `m_` prefix for instance data members.  Also, remember to register for save states in `machine_start`.

File: src/mame/drivers/kinst.cpp [hunk 4]
```diff
@@ -881,4 +908,4 @@ void kinst_state::init_kinst2()
 
 GAME( 1994, kinst,    0,      kinst, kinst,  kinst_state, init_kinst,  ROT0, "Rare", "Killer Instinct", MACHINE_SUPPORTS_SAVE )
 GAME( 1995, kinst2,   0,      kinst, kinst2, kinst_state, init_kinst2, ROT0, "Rare", "Killer Instinct 2", MACHINE_SUPPORTS_SAVE )
-GAME( 1995, kinst2uk, kinst2, kinst, kinst2, kinst_state, init_kinst2, ROT0, "Rare", "Killer Instinct 2 (Upgrade kit)", MACHINE_NOT_WORKING | MACHINE_SUPPORTS_SAVE )
+GAME( 1995, kinst2uk, kinst2, kinst, kinst2, kinst_state, init_kinst, ROT0, "Rare", "Killer Instinct 2 (Upgrade kit)", MACHINE_SUPPORTS_SAVE )
```
Comment (cuavas) [inline src/mame/drivers/kinst.cpp:911]: Please maintain tabulation.

**Review by cuavas — CHANGES_REQUESTED** [review]
Please address the issues and give the pull request a descriptive title.

**Comment by cuavas** [comment]
I implemented it quickly and cleaned up some other stuff in the driver: 986a98c51d4fe993a0a0a045edcd9ac8c315f0ec

I give it a quick test, and it seemed to work fine.  Let me know if it looks like I missed something.  I’m still not sure about zeroing the registers in `machine_reset` – does the CPLD really have access to that signal?

**Comment by rb6502** [comment]
ATA does have a reset line, so it's possible.

**Comment by pdaderko** [comment]
Thanks to everyone for the help, and taking it (somewhat) easy on me.  I've learned a lot from this, and if there's a next time, hopefully it'll go smoother because of this.

> I give it a quick test, and it seemed to work fine. Let me know if it looks like I missed something.

It looks correct to me.

> I’m still not sure about zeroing the registers in `machine_reset` – does the CPLD really have access to that signal?

The CPLD gets the reset line from the IDE connector, which does get asserted at reset, though I remember questioning whether it actually reset or not when I first started looking at this.  I started power cycling every test because the power-on reset was definitely consistent.  I'm tweaking my test ROM and hooking it back to the logic analyzer to find out for sure... I'll post back once I have an answer.

**Comment by cuavas** [comment]
>Is this described somewhere? It seems that the MAMEdev wiki is down... I found a cached page talking about it, but it looks out of date ( https://wiki.mamedev.org/index.php/Save_State_Fundamentals ), unless I'm just not looking in the right spot to find anything described on that page.

Unfortunately the wiki is down right now, but it should be coming back sooner or later.  Unfortunately wikis rely on someone actually keeping stuff up-to-date, so the quality of the documentation there is, shall we say, rather inconsistent.  MAME has gone through a lot of major architectural changes, and not everything has been kept up to date.

>Side note - I find there's a lot of learning curve because of all the MAME specific references, and I don't know an easy way to see what the code actually does (like "go to definition" like I'd do in an Eclipse based project). Is there auto-generated code documentation somewhere that I'm missing?

The “go to definition” feature does work in several editors/IDEs, including vim with ctags, Visual Studio Code, and Visual Studio.  It works a lot better than it used to, now that we don’t hide so many things under layers of macros.  I’m not sure whether it works with Eclipse or not.

There’s some up-to-date documentation aimed at developers here: https://docs.mamedev.org/techspecs/

Several core components are covered reasonably well, but it’s definitely very incomplete.  It’s also more of a reference than an introduction/tutorial for the most part.  There were some tutorials on the wiki, but they were never kept up-to-date and ended up being more hindrance than help.

Some key classes do have Doxygen format documentation that some IDEs will pick up automatically.  Some examples:
* The IoC container lifecycle members of `device_t`: https://github.com/mamedev/mame/blob/mame0235/src/emu/device.h#L668
* The lifecycle members of `device_interface`: https://github.com/mamedev/mame/blob/mame0235/src/emu/device.h#L873
* Pretty much everything in `devfind.h` (also documented [here](https://docs.mamedev.org/techspecs/object_finders.html)): https://github.com/mamedev/mame/blob/mame0235/src/emu/devfind.h
* The bit shuffling helper functions: https://github.com/mamedev/mame/blob/mame0235/src/lib/util/coretmpl.h#L566

Actually building the Doxygen documentation is broken at the moment because the Bootstrapped theme is incompatible with the latest version of Doxygen (newer versions of Doxygen generate output that uses jQuery SmartMenus, which the Bootstrapped theme doesn’t support).  You can get it to work if you comment out a few lines in doxygen/doxygen.config and then doing `make doxygen` (output ends up in output-doc).  You’ll end up with a really large number of pages listing class members, relatively few of which have complete documentation.

The Doxygen documentation is admittedly very incomplete, for the simple reason that developers generally don’t want to put time into writing it.  It’s more “fun” to get on with the next development task, or there’s something else broken that needs fixing, and so on.

>Some things are easy to find by searching, but others are used so frequently that finding the definition based on a keyword search is nearly impossible (e.g. the definition of GAME() ).

Well, Visual Studio Code, Visual Studio and vim all found the definition of the `GAME` macro for me, so I’m not sure what to say about that.  For what it’s worth, it lives here: https://github.com/mamedev/mame/blob/mame0235/src/emu/gamedrv.h#L167

The trouble is, even if you look at the definition, it’s scary and not at all obvious what it does, and that’s coming from the person who wrote the current implementation of it.  There’s a wiki page that briefly explains what the parameters are (isn’t it great having the wiki down), but it’s still not immediately obvious how you’re supposed to use it.

A lot of the time, people work on MAME by finding something vaguely similar to what they want to do, and then trying to modify it until it works.  This is admittedly a lot easier if you’re familiar with a lot of MAME, because you’re more likely to know that something similar exists and where to find it.

I get it, MAME is poorly documented and the learning curve is steep.  I *am* trying to make it more approachable, including trying to make the code look more like idiomatic C++, documenting core functionality as I work on it, etc.  But given the size, scope and age of the project, I think it’s always going to be a bit daunting when getting started.

But all that aside, thanks for the contribution.  It’s great to have the upgrade kit protection documented.

**Comment by pdaderko** [comment]
> I’m still not sure about zeroing the registers in `machine_reset` – does the CPLD really have access to that signal?

I tested this, and can confirm it does NOT reset the register values when the board is reset warm (pressing the reset button on the PCB), even though the IDE connector's /RESET line does toggle, which is connected to the CPLD.


Regarding documentation... I totally get that documentation for a large and very mature project with many contributors run by volunteers will certainly be lacking, because yeah, doing real code is more fun.  Proper Doxygen generated documentation is a thing of beauty though, and I'm definitely spoiled by (for example) some Silicon Labs documentation, like: https://siliconlabs.github.io/Gecko_SDK_Doc/efm32gg/html/group__GPIO.html .  Human readable descriptions, clickable links take you to enumerations, definitions, etc.  And you can click from inside the code to go to its documentation.  But of course that's API code for commercial products.

And yes, I probably should have installed Visual Studio or similar to navigate the code better.  I'm simply editing (with Notepad2), since I only needed to mod one file - then using the prebuilt MAME Windows build tools.  I see github supports better code navigation (things like Jump to definition) for certain languages, but C++ isn't one of them.

Regarding coding styles, etc. I looked for the "contributing guidelines", but it's just a blank page: https://github.com/mamedev/mame/blob/master/docs/CONTRIBUTING.md .  So, I tried to make my code consistent with the rest of the code in the file, but obviously I missed quite a few things.  And for example, I changed all my comments from // to /**/ because the rest of the code had it that way, but after seeing your changes, I guess that wasn't the preferred style anyway.  So, if a coding style page doesn't already exist, having one might reduce some of the simple issues.

Anyway, thanks again.

**Comment by happppp** [comment]
Merged manually by cuavas, closing PR

**Pull request closed without merging**